wgengine/tsdns/map.go

```diff
@@ -0,0 +1,66 @@
+// Copyright (c) 2020 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package tsdns
+
+import (
+	"fmt"
+	"strings"
+
+	"inet.af/netaddr"
+)
+
+// Map is all the data Resolver needs to resolve DNS queries within the Tailscale network.
+type Map struct {
+	// domainToIP is a mapping of Tailscale domains to their IP addresses.
+	// For example, monitoring.tailscale.us -> 100.64.0.1.
+	domainToIP map[string]netaddr.IP
+}
+
+// NewMap returns a new Map with domain to address mapping given by domainToIP.
+func NewMap(domainToIP map[string]netaddr.IP) *Map {
+	return &Map{domainToIP: domainToIP}
+}
+
+func printSingleDomain(buf *strings.Builder, domain string, ip netaddr.IP) {
+	// Output width is exactly 80 columns.
+	fmt.Fprintf(buf, "%-63s %15s\n", domain, ip)
```

Comment:
the %63 seems a bit wide for the normal case of short names. and %15s won't deal well with IPv6 anyway.

Maybe just "%s\t%s\n" ?

Reply:
That would work, but some domain names end up much longer than others. On my personal account, I get a 53-character name without trying (`dmytros-imac-pro-1.dm-shynk.gmail.com.b.tailscale.net`), so I thought this was a reasonable width:
```
5.9M/20.8M tsdns: map diff:
+admins-imac-pro-1.dm-shynk.gmail.com.b.tailscale.net             100.125.82.109
+admins-imac-pro.dm-shynk.gmail.com.b.tailscale.net                100.115.22.60
+hello-ipn-dev.tailscale.com.b.tailscale.net                     100.101.102.103
+mi-a1-1.dm-shynk.gmail.com.b.tailscale.net                        100.69.189.55
+oberon.dm-shynk.gmail.com.b.tailscale.net                         100.108.6.127
+ubuntu16.dm-shynk.gmail.com.b.tailscale.net                      100.106.215.35
+ubuntu20.dm-shynk.gmail.com.b.tailscale.net                       100.65.242.95
+void-2.dm-shynk.gmail.com.b.tailscale.net                           100.99.3.26
+void.dm-shynk.gmail.com.b.tailscale.net                          100.81.169.109
```

Reply:
On second look, it doesn't look _that_ back. Changed.